# Guided Capstone Step 5. Modeling 

**The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this guided capstone we are going to revisit many of the actions we took in the previous guided capstone. This gives you the opportunity to practice the code you wrote to solve the questions in step 4. 

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [56]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
os.listdir()

['.conda',
 '.condarc',
 '.gitconfig',
 '.idea',
 '.idlerc',
 '.ipynb_checkpoints',
 '.ipython',
 '.jupyter',
 '.matplotlib',
 '01_first_notebook.ipynb',
 '1587423569_GuidedCapstoneStep6-AnswerKey.ipynb.zip',
 '3D Objects',
 'Anaconda3',
 'AppData',
 'Application Data',
 'Contacts',
 'Cookies',
 'Desktop',
 'Documents',
 'Downloads',
 'Dropbox',
 'Favorites',
 'folder',
 'Google Drive',
 'GuidedCapstoneStep3-AnswerKeyHL.ipynb',
 'GuidedCapstoneStep3-AnswerKeyHL0.ipynb',
 'GuidedCapstoneStep3-AnswerKeyHL_2.ipynb',
 'GuidedCapstoneStep3-AnswerKeyHL_3.ipynb',
 'GuidedCapstoneStep3HL.ipynb',
 'GuidedCapstoneStep3HL_test.ipynb',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'GuidedCapstoneStep4-AnswerKey.ipynb',
 'GuidedCapstoneStep5-AnswerKey.ipynb',
 'GuidedCapstoneStep6-AnswerKey.ipynb',
 'GuidedCapstoneStep6HL.ipynb',
 'GuidedCapstone_Step2HL .ipynb',
 'IntelGraphicsProfiles',
 'Links',
 'Local Settings',
 'London Districts Challenge _Ch.#4.ipynb',
 'MicrosoftEdgeBackups',
 'Music',
 'My D

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [57]:
file="data/step3_output.csv"
df=pd.read_csv(r'C:\Users\zsoltani\Desktop\Zohreh Training and Personal files\Python class\Chapter 6\data\step3_output.csv')
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         2  
1            80.0         0  
2             0.0         0  
3             0.0         0  
4             0.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

In [78]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(df.shape)
df.head()

(164, 60)


Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3               Donner Ski Ranch         8012            750      0   
4                  June Mountain        10090           2590      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          0     0       1       5  ...             0   
4        0.0          0          2     0       0       4  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [79]:
#scale data
from sklearn import preprocessing
X=df.drop(['Name','AdultWeekend'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [80]:
from sklearn.model_selection import train_test_split
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn fit the model on your training dataset.</font>**

#### Model 1

In [81]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [82]:
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed.

In [83]:
explained_variance_score(y_test, y_pred)

-2.7194056779723954e+23

In [84]:
mean_absolute_error(y_test, y_pred)

1082751857661.931

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [85]:
lm.intercept_

270687964471.02106

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for `AdultWeekend` resulting value. Also, because we took the time to scale our x values in the training data we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [86]:
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
fastSixes      7.076941e+13
fastEight      3.887508e+13
total_chairs   1.447775e+13
New York       1.420790e+13
Michigan       1.279699e+13
New Hampshire  1.110481e+13
Pennsylvania   1.062256e+13
Wisconsin      1.011011e+13
California     9.562605e+12
Minnesota      9.562605e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [87]:
file="data/step3_output.csv"
df=pd.read_csv(r'C:\Users\zsoltani\Desktop\Zohreh Training and Personal files\Python class\Chapter 6\data\step3_output.csv')
X=df.drop(['AdultWeekend','Name','state'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
mean_absolute_error(y_test, y_pred)

0.6051137738602255


6.837916672506848

In [88]:
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday         12.041115
summit_elev           2.870288
vertical_drop         2.741865
daysOpenLastYear      2.029302
quad                  1.760005
SkiableTerrain_ac     1.576850
projectedDaysOpen     1.424838
clusters              1.312052
averageSnowfall       1.260074
Runs                  1.196191

**<font color='teal'> When reviewing our new model coefficients we see `summit_elev` is now in the number two spot, this also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [89]:
file="data/step3_output.csv"
df=pd.read_csv(r'C:\Users\zsoltani\Desktop\Zohreh Training and Personal files\Python class\Chapter 6\data\step3_output.csv')
X=df.drop(['Name','AdultWeekend','state','summit_elev'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
mean_absolute_error(y_test, y_pred)

0.598704311667656


6.903861802988329

In [90]:
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday         12.084045
vertical_drop         2.806302
SkiableTerrain_ac     1.915104
daysOpenLastYear      1.782833
quad                  1.611917
projectedDaysOpen     1.388388
averageSnowfall       1.190371
LongestRun_mi         1.165913
clusters              1.102006
double                1.084137

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell and you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | |  |-|
| Model 2. |0.605| 6.84 |'state'|
| Model 3. |0.599| 6.93|'state','summit_elev','base_elev'|

Model Selection: